pkg/cloudprovider/nodecontroller/nodecontroller.go

```diff
@@ -648,15 +648,25 @@ func (nc *NodeController) getCloudNodesWithSpec() (*api.NodeList, error) {
 	return result, nil
 }
 
+func getHostFieldLabel(apiVersion string) string {
+	switch apiVersion {
+	case "v1beta1", "v1beta2":
+		return "DesiredState.Host"
+	default:
+		return "spec.host"
+	}
+}
+
 // deletePods will delete all pods from master running on given node.
 func (nc *NodeController) deletePods(nodeID string) error {
 	glog.V(2).Infof("Delete all pods from %v", nodeID)
-	// TODO: We don't yet have field selectors from client, see issue #1362.
-	pods, err := nc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+	pods, err := nc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(),
+		fields.OneTermEqualSelector(getHostFieldLabel(nc.kubeClient.APIVersion()), nodeID))
```

Comment:
There is not need to introduce the APIVersion mechanism.
Basically, what you should use is "client.PodHost" - see e.g.:
https://github.com/GoogleCloudPlatform/kubernetes/blob/d0288f714328e82f54f782bfa610cb3cb285c17f/pkg/kubelet/config/apiserver.go#L30
And this should be handled correctly by our infrastructure - see:
https://github.com/GoogleCloudPlatform/kubernetes/blob/d0288f714328e82f54f782bfa610cb3cb285c17f/pkg/client/request.go#L295